### ISSUE TITLE
Possible bug writing pattern to tags file

### ISSUE THREAD

**Issue by ffes** [universal-ctags/ctags#373]
While testing the rest.c parser from Geany, I run into a something strange. That parser has a bug and adds the underline as pattern. There seems to be an easy fix by adding one line to the makeRestTag() `e.pattern = e.name;`

But with this fix when the tags line is written to the file the `/^` and `$/` are not added. This is because at https://github.com/universal-ctags/ctags/blob/master/main/entry.c#L905  the pattern is just written to the file. This seems to be a bug to me. I think this test should be in writePatternEntry() just above that.

So before I open a PR for that I would like your opinion on this. Maybe I just overlooked something.

This issue needs to be solved before I can add the reStructuredText parser and with that add nestlevel.[ch] aka #315

**Comment by b4n** [comment]
> While testing the rest.c parser from Geany, I run into a something strange. That parser has a bug and adds the underline as pattern. There seems to be an easy fix by adding one line to the makeRestTag() `e.pattern = e.name;`

I didn't look into it, but I guess the real fix would be to have the proper line/position information on the tag, e.g. the line with the title and not the one with the underline.

> But with this fix when the tags line is written to the file the `/^` and `$/` are not added. […] This seems to be a bug to me.

While it might be sensible to change this as you said, AFAIK *no* parsers currently set `entry->pattern`, so it might just be considered an internal detail, and parsers *should not* be setting it.  Also if they really wanted to do it, it might make sense for them to chose not to use completely anchored patterns (imagine e.g. minified CSS or JS).

**Comment by ffes** [comment]
> I didn't look into it, but I guess the real fix would be to have the proper line/position
> information on the tag, e.g. the line with the title and not the one with the underline.

Yes, storing the file position was not that difficult either. I solved the problem with that. I will prepare a PR for reStructuredText.

> While it might be sensible to change this as you said, AFAIK no parsers currently
> set `entry->pattern`, so it might just be considered an internal detail, and parsers
> should not be setting it.

I leave it for now as it is. I will investigate the use of the pattern member of the struct some time later.

> Also if they really wanted to do it, it might make sense for
> them to chose not to use completely anchored patterns (imagine e.g. minified CSS or JS).

Minified JS and CSS result in quite useless tags anyway IMHO. In my tags plugin for Notepad++ I ignore tags with patterns longer then 1024 chars in JS and CSS.

**Comment by masatake** [comment]
I've tried rst parser. It prints the line number and the pattern for a tag well.
If the issue is reproduced, please reopen this.
